src/tabs/console/ConsoleAutocomplete.tsx

```diff
@@ -0,0 +1,70 @@
+// Tab completion engine for autocompleting commands. Inspired by the TabCompletionEngine class
+// for the Terasology in-game console.
+
+export class ConsoleAutocomplete {
+
+  // autocomplete the command.
+  public static complete(command: string): string {
+    this.previousCommands.push(command);
+    // if command was not entered, display some help text
+    if (command.length <= 0) {
+      return "Type 'help' to list all commands.";
+    }
+    // if no commands were previously entered, find a list of commands
+    if (this.previousCommands.length === 1) {
+      this.findMatches(command);
+    }
+    // if matches were found, do nothing to the command
+    if (this.matches.length === 0) {
+      return command;
+    } else if (this.matches.length === 1) {
+      return this.matches[0];
+    }
+    // If the last possible match has already been iterated through, reset the index to the first location.
+    // The match index must be offset because the length function returns 1 with one element, but its index
+    // is zero.
+    if (this.matchIndex + 1 === this.matches.length) {
+      this.matchIndex = -1;
+    }
+    // If the first command entered had more than one match, cycle throught the matches.
+    // previousCommands is always at least size 1.
+    if (this.previousCommands.indexOf(this.previousCommands[0]) !== -1 && this.previousCommands.length !== 1) {
+      ++this.matchIndex;
+    }
+    return this.matches[this.matchIndex];
+  }
+
+  // reset the state of autocompletion
+  public static clear() {
+    this.matches = [];
+    this.matchIndex = 0;
+    this.previousCommands = [];
+  }
+
+  // set the list of commands provided by the backend
+  public static setCommandList(commands: string[]) {
+    if (this.commandList.length === 0) {
+      this.commandList = commands;
+    }
+  }
+
+  // the matches currently found
+  private static matches: string[] = [];
```

Comment:
I'd put the fields at the top of the class, before all the methods. Did you put them here because `tslint` would otherwise complain? I remember some linter tool trying to enforce a certain order of all the class members but I don't remember if it was in this project, if it is maybe it's time to do some tweaks at tslint settings (if you put the fields here for this or a similar reason I'll merge anyway then try to fix the tool's configuration myself)

Reply:
yes, it causes tslint to complain. It's an easy fix in tslint.json though.